Cloudnative

gizas opened the pull request:
The new track will be called cloudnative.

Aim is this track to provide:
- Corpora data available to Gooogle Cloud that include data from a real kubenretes cluster
- Test cases performed on specific indices
- Data and mappings will have to be extracted from rally `create-track` command

Track tested locally with multiple successful runs:



-----
Example:

❯ esrally race --distribution-version=8.3.0 --track-path=/Users/andreasgkizas/.rally/benchmarks/tracks/cloudnative --runtime-jdk=bundled --kill-running-processes

```

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Race id is [14eb7c6a-37d1-4916-bd60-d9b2e35cae2c]
[INFO] Preparing for race ...
[INFO] Racing on track [cloudnative], challenge [append-no-conflicts-index-only] and car ['defaults'] with version [8.3.0].

Running delete-index                                                           [100% done]
Running create-index                                                           [100% done]
Running check-cluster-health                                                   [100% done]
Running index-stats                                                            [100% done]
Running index-append                                                           [100% done]
Running index-update                                                           [100% done]
Running node-stats                                                             [100% done]
Running query_all                                                              [100% done]
Running query_aggregation                                                      [100% done]

------------------------------------------------------
    _______             __   _____
   / ____(_)___  ____ _/ /  / ___/_________  ________
  / /_  / / __ \/ __ `/ /   \__ \/ ___/ __ \/ ___/ _ \
 / __/ / / / / / /_/ / /   ___/ / /__/ /_/ / /  /  __/
/_/   /_/_/ /_/\__,_/_/   /____/\___/\____/_/   \___/
------------------------------------------------------

|                         Metric |              Task |        Value |   Unit |
|-------------------------------:|------------------:|-------------:|-------:|
|        Total Young Gen GC time |                   |     70.82    |      s |
|       Total Young Gen GC count |                   |  11655       |        |
|          Total Old Gen GC time |                   |      0       |      s |
|         Total Old Gen GC count |                   |      0       |        |
|    Total Ingest Pipeline count |                   |      0       |        |
|     Total Ingest Pipeline time |                   |      0       |      s |
|   Total Ingest Pipeline failed |                   |      0       |        |
|                 Min Throughput |       index-stats |     90       |  ops/s |
|                Mean Throughput |       index-stats |     90       |  ops/s |
|              Median Throughput |       index-stats |     90       |  ops/s |
|                 Max Throughput |       index-stats |     90.01    |  ops/s |
|        50th percentile latency |       index-stats |      3.13722 |     ms |
|        90th percentile latency |       index-stats |      3.82281 |     ms |
|        99th percentile latency |       index-stats |      4.82432 |     ms |
|      99.9th percentile latency |       index-stats |     18.2555  |     ms |
|       100th percentile latency |       index-stats |     28.2685  |     ms |
|   50th percentile service time |       index-stats |      1.77596 |     ms |
|   90th percentile service time |       index-stats |      2.28356 |     ms |
|   99th percentile service time |       index-stats |      2.93292 |     ms |
| 99.9th percentile service time |       index-stats |      5.78041 |     ms |
|  100th percentile service time |       index-stats |     27.8386  |     ms |
|                     error rate |       index-stats |      0       |      % |
|                 Min Throughput |      index-append |  23470.3     | docs/s |
|                Mean Throughput |      index-append |  29962       | docs/s |
|              Median Throughput |      index-append |  30635.2     | docs/s |
|                 Max Throughput |      index-append |  32090.7     | docs/s |
|        50th percentile latency |      index-append |    936.052   |     ms |
|        90th percentile latency |      index-append |   1793.94    |     ms |
|        99th percentile latency |      index-append |   3275.98    |     ms |
|      99.9th percentile latency |      index-append |   5131.17    |     ms |
|       100th percentile latency |      index-append |   6498.13    |     ms |
|   50th percentile service time |      index-append |    936.052   |     ms |
|   90th percentile service time |      index-append |   1793.94    |     ms |
|   99th percentile service time |      index-append |   3275.98    |     ms |
| 99.9th percentile service time |      index-append |   5131.17    |     ms |
|  100th percentile service time |      index-append |   6498.13    |     ms |
|                     error rate |      index-append |      7.5     |      % |
|                 Min Throughput |      index-update |  13479.7     | docs/s |
|                Mean Throughput |      index-update |  17543       | docs/s |
|              Median Throughput |      index-update |  17890.6     | docs/s |
|                 Max Throughput |      index-update |  18843.6     | docs/s |
|        50th percentile latency |      index-update |   1063.36    |     ms |
|        90th percentile latency |      index-update |   2586.87    |     ms |
|        99th percentile latency |      index-update |   5334.6     |     ms |
|      99.9th percentile latency |      index-update |   7739.44    |     ms |
|       100th percentile latency |      index-update |  10861.5     |     ms |
|   50th percentile service time |      index-update |   1063.36    |     ms |
|   90th percentile service time |      index-update |   2586.87    |     ms |
|   99th percentile service time |      index-update |   5334.6     |     ms |
| 99.9th percentile service time |      index-update |   7739.44    |     ms |
|  100th percentile service time |      index-update |  10861.5     |     ms |
|                     error rate |      index-update |     36.35    |      % |
|                 Min Throughput |        node-stats |     89.04    |  ops/s |
|                Mean Throughput |        node-stats |     89.63    |  ops/s |
|              Median Throughput |        node-stats |     89.72    |  ops/s |
|                 Max Throughput |        node-stats |     89.84    |  ops/s |
|        50th percentile latency |        node-stats |      3.19215 |     ms |
|        90th percentile latency |        node-stats |      5.22003 |     ms |
|        99th percentile latency |        node-stats |     22.46    |     ms |
|      99.9th percentile latency |        node-stats |     69.6661  |     ms |
|       100th percentile latency |        node-stats |     74.6041  |     ms |
|   50th percentile service time |        node-stats |      2.04546 |     ms |
|   90th percentile service time |        node-stats |      3.96094 |     ms |
|   99th percentile service time |        node-stats |      6.62663 |     ms |
| 99.9th percentile service time |        node-stats |     17.6867  |     ms |
|  100th percentile service time |        node-stats |     73.3922  |     ms |
|                     error rate |        node-stats |      0       |      % |
|                 Min Throughput |         query_all |     45.83    |  ops/s |
|                Mean Throughput |         query_all |     47.48    |  ops/s |
|              Median Throughput |         query_all |     47.67    |  ops/s |
|                 Max Throughput |         query_all |     48.38    |  ops/s |
|        50th percentile latency |         query_all |      7.16425 |     ms |
|        90th percentile latency |         query_all |      9.79979 |     ms |
|        99th percentile latency |         query_all |     21.3653  |     ms |
|      99.9th percentile latency |         query_all |     52.1686  |     ms |
|       100th percentile latency |         query_all |     66.1326  |     ms |
|   50th percentile service time |         query_all |      5.6355  |     ms |
|   90th percentile service time |         query_all |      8.09916 |     ms |
|   99th percentile service time |         query_all |     15.6535  |     ms |
| 99.9th percentile service time |         query_all |     30.4881  |     ms |
|  100th percentile service time |         query_all |     64.2519  |     ms |
|                     error rate |         query_all |      0       |      % |
|                 Min Throughput | query_aggregation |      2.62    |  ops/s |
|                Mean Throughput | query_aggregation |      2.64    |  ops/s |
|              Median Throughput | query_aggregation |      2.64    |  ops/s |
|                 Max Throughput | query_aggregation |      2.65    |  ops/s |
|        50th percentile latency | query_aggregation | 359094       |     ms |
|        90th percentile latency | query_aggregation | 500608       |     ms |
|        99th percentile latency | query_aggregation | 532466       |     ms |
|      99.9th percentile latency | query_aggregation | 535633       |     ms |
|       100th percentile latency | query_aggregation | 536002       |     ms |
|   50th percentile service time | query_aggregation |    372.123   |     ms |
|   90th percentile service time | query_aggregation |    376.646   |     ms |
|   99th percentile service time | query_aggregation |    419.581   |     ms |
| 99.9th percentile service time | query_aggregation |    785.207   |     ms |
|  100th percentile service time | query_aggregation |    815.442   |     ms |
|                     error rate | query_aggregation |      0       |      % |

[WARNING] Error rate is 7.5 for operation 'index-append'. Please check the logs.
[WARNING] Error rate is 36.35 for operation 'index-update'. Please check the logs.

----------------------------------
[INFO] SUCCESS (took 2485 seconds)
----------------------------------
```

DJRickyB reviewed (changes requested):
You'll want to ensure the track succeeds. We have explicit failures, e.g.

```
[WARNING] Error rate is 7.5 for operation 'index-append'. Please check the logs.
[WARNING] Error rate is 36.35 for operation 'index-update'. Please check the logs.
```

But also failures of tuning/configuration. See here:
|                         Metric |              Task |        Value |   Unit |
|-------------------------------:|------------------:|-------------:|-------:|
|        50th percentile latency | query_aggregation | 359094       |     ms |
|        90th percentile latency | query_aggregation | 500608       |     ms |
|        99th percentile latency | query_aggregation | 532466       |     ms |
|      99.9th percentile latency | query_aggregation | 535633       |     ms |
|       100th percentile latency | query_aggregation | 536002       |     ms |
|   50th percentile service time | query_aggregation |    372.123   |     ms |
|   90th percentile service time | query_aggregation |    376.646   |     ms |
|   99th percentile service time | query_aggregation |    419.581   |     ms |
| 99.9th percentile service time | query_aggregation |    785.207   |     ms |
|  100th percentile service time | query_aggregation |    815.442   |     ms |
```

'latency' should only be slightly higher than `service_time` in a throttled and tuned configuration. See more: https://esrally.readthedocs.io/en/stable/faq.html#what-do-latency-and-service-time-mean-and-how-do-they-relate-to-the-took-field-that-elasticsearch-returns

The above certainly need to be addressed. The below are more philosophical:

I'm also concerned about the maintainability of the current track code. It seems whenever it gets updated it will require essentially starting from scratch each time. In addition, due to the current process of extracting from backing indices, it may miss the addition of ingest pipelines (it already is missing the inclusion of the Fleet final pipeline in throughput calculations).

Also, I'm curious why the benchmark includes `node-stats` and `index-stats` operations. Do you perhaps mean to get their data, via `--telemetry` instead? More info: https://esrally.readthedocs.io/en/stable/telemetry.html

gizas commented:
After several efforts track tuned with no errors in running. See sample in file:
[cloudnative_run.txt](https://github.com/elastic/rally-tracks/files/9131592/cloudnative_run.txt)

Added also Readme update information on how to use telemetry suggestions (Thank you @DJRickyB )

For the create-track command it is known that no ingest pipelines are being extracted. Maybe full extraction to be added later in rally or on elastic-package dump as you pointed https://github.com/elastic/rally-tracks/issues/280

But for now I think is out of scope to include this scenario.


Finally  node-stats and index-stats were just added as a general benchmark for the track. No harm to include them and run pretty quickly.

Finally the `query_aggregation` was a specific query that we know that produces a lot of delay and considered expensive as from relevant previous bug analysis [here](https://github.com/elastic/k8s-integration-infra#helper-es_bench-script-tsdb-use-case-only)

gizas commented:
[cloudnative_run_synthetics.txt](https://github.com/elastic/rally-tracks/files/9134630/cloudnative_run_synthetics.txt)

A successful run with synthetics enabled

DJRickyB reviewed (changes requested):
I would be curious about stated goals for this track that I've missed, it seems like it is in service to a very specific one-off effort for evaluating synthetic `_source` for observability integrations, and doesn't necessarily need to be in `rally-tracks`. My concern about committing it here is the implication that it should be maintained, regularly run, or serve as an example of how to create a Rally benchmark for datastream-oriented use cases, all of which I think are not the point.

Alternatively, I've tried to leave more tactical feedback in-line. Also; per the attached results:

```
[WARNING] Error rate is 7.03 for operation 'index-append'. Please check the logs.
[WARNING] Error rate is 6.96 for operation 'index-update'. Please check the logs.
```

This still likely should not be happening in order to get meaningful results from the indexing tasks, though my concern (in that case) about not having ingest pipelines from Fleet still stands

ruflin commented:
@gizas What is the plan with this PR? Would be good to get this over the line otherwise it gets quickly outdated and cannot be run by others.

gizas commented:
@ruflin I am not even sure as how we need to proceed from this point on. 

@DJRickyB below commented: 
> I would be curious about stated goals for this track that I've missed, it seems like it is in service to a very specific one-off effort for evaluating synthetic _source for observability integrations, and doesn't necessarily need to be in rally-tracks.

So on one hand I agree that this work is not needed to be merged to rally-tracks at all, or if we want we can remove extra work for synthetics and keep it as simple as possible with as little tests as we can. But again every next time that someone will need to reuse it for an extra feature he might need to update the indexes and mappings.

Nevertheless, after next week we can have a closer look and decide what we can do

gizas commented on cloudnative/files.txt:
```diff
@@ -0,0 +1,74 @@
+.ds-logs-elastic_agent.metricbeat-default-2022.06.20-000001-documents-1k.json.bz2
+.ds-logs-elastic_agent.metricbeat-default-2022.06.20-000001-documents.json.bz2
+.ds-logs-kubernetes.container_logs-default-2022.06.20-000001-documents-1k.json.bz2
+.ds-logs-kubernet
```
fixed

gizas commented:
@DJRickyB and @ruflin I have made a final push [f3dbf7b](https://github.com/elastic/rally-tracks/pull/278/commits/f3dbf7b447837c54f52363576bcacbb309977f84) 

See results for 8.4.0 run:
[report8.4.0.md](https://github.com/elastic/rally-tracks/files/9495586/report8.4.0.md)

I believe we are ready to merge, addressed above comments, updated queries.
@DJRickyB to answer your concerns @elastic/obs-cloudnative-monitoring team will monitor this track and we make efforts to run this once a month

dliappis reviewed (commented):
Left a few small comments, but also discussed offline with @gizas about the goals, as well as intended plans for keeping it up to date. The understanding is:

1. For now, the intention is to run this occasionally -- manually triggered -- targetting released versions of the stack.

2. There is a strong desire to keep it up to date (with an automated way) in terms of mappings as well as corpora. I mentioned the [elastic-package dump](https://github.com/elastic/elastic-package/issues/601#issuecomment-1036305966) workflow and it became obvious that there is a lot of room for synergies with the work that @michaelbaamonde is doing to "align the workloads with solution releases".

Based on the above there is no concrete advantage to store this in the repo of the default rally-tracks, esp. given the need to stay on top of backports for branches, but instead keep it in a different repo: the only advantage we could see keeping it here, is easy access from use cases via our orchestration tool (`esbench`). The last point, however, is is a very small thing to address vs having public users scratching heads by potentially hitting errors with this track due to version support gaps.

Once the plumbing to keep this track up to date (mappings, corpora etc.) has been put in place I believe it'll be a very useful addition here, and probably a good candidate for a regularly scheduled benchmark in https://elasticsearch-benchmarks.elastic.co.

gizas commented on cloudnative/README.md:
````diff
@@ -0,0 +1,173 @@
+Corpora data of the specific data with their respective mappings were extracted with following command:
+
+```bash
+esrally create-track --track=test --target-hosts=test.rally.gizas.elastic-observability-ocp.elastic.dev:9200 --client-options="use_ssl:true,verify_certs:false,ca_certs:false,basic_auth_user:'elastic',basic_auth_password:'C7GfaE950v74IwPZq6'" --datastreams="metrics-*, logs-*" --output-path=~/tracks
````
done

gizas commented on cloudnative/README.md:
```diff
@@ -0,0 +1,173 @@
+
+### Fetching new data
+
+This data comes from a real cluster that has been deployed with Terraform of [k8s-integration-infra](https://github.com/elastic/k8s-integration-infra/tree/main/infra)
```
Indeed added a comment and also gave u access

gizas commented on cloudnative/README.md:
```diff
@@ -0,0 +1,173 @@
+
+Extraction of data-streams and indexes from line Elastic Cloud was done with the help of rally `create-index`
+
+Modifications:
```
Updated

pquentin commented:
@gizas What is the status of this pull request?

gizas commented:
@pquentin closing this PR request.
The backup of cloudnative tracked moved under https://github.com/elastic/k8s-integration-infra/tree/main/cloudnative

alvarolobato commented:
@gizas when is this track actually being executed? I would like to make sure we have a set of performance tests targeted to k8s support in TSDB nightly testing, is that this track?

cc @ruflin
